src/attribute.cpp

```diff
@@ -27,6 +43,172 @@ Integer::Integer(uint8_t type, const uint8_t* data, size_t size)
             + data[3];
 }
 
+Integer::Integer(uint8_t type, uint32_t value)
+    : Attribute(type),
+      m_value(value)
+{
+}
+
+std::vector<uint8_t> Integer::toVector(const std::string& secret, const std::array<uint8_t, 16>& auth) const
+{
+    std::vector<uint8_t> attribute(6);
+    attribute[0] = type();
+    attribute[1] = 6;
+    attribute[2] = m_value / (1 << 24);
+    attribute[3] = (m_value / (1 << 16)) % 256;
+    attribute[4] = (m_value / (1 << 8)) % 256;
+    attribute[5] = m_value % 256;
+    return attribute;
+}
+
+IpAddress::IpAddress(uint8_t type, const uint8_t* data, size_t size)
+        : Attribute(type)
+{
+    if (size != 4)
+        throw std::runtime_error("Invalid integer attribute size. Should be 4, actual size is " + std::to_string(size));
+
+    for (size_t i = 0; i < size; ++i)
+        m_value[i] = data[i];
+}
+
+IpAddress::IpAddress(uint8_t type, std::array<uint8_t, 4> address)
+    : Attribute(type),
+      m_value(address)
+{
+}
+
+std::vector<uint8_t> IpAddress::toVector(const std::string& secret, const std::array<uint8_t, 16>& auth) const
+{
+    std::vector<uint8_t> attribute(6);
+
+    attribute[0] = type();
+    attribute[1] = attribute.size();
+    for (size_t i = 0; i < m_value.size(); ++i)
+        attribute[i + 2] = m_value[i];
+
+    return attribute;
+}
+
+Encrypted::Encrypted(uint8_t type, const uint8_t* data, size_t size, std::string secret, std::array<uint8_t, 16> auth)
+        : Attribute(type)
+{
+    if (size > 128)
+        throw std::runtime_error("Invalid encrypted attribute size. Should be max 128 bytes, actual size is " + std::to_string(size));
+
+    std::vector<uint8_t> value(size);
+
+    size_t j = 16;
```

Comment:
Consider the following:
![Encryption/decryption diagram](https://user-images.githubusercontent.com/7895530/193596402-bebe0623-8daf-43d4-854e-fd82743d8f19.png)

This diagram illustrates encryption and decryption process. It helps to exctract key elements for the implementation - input data, cycle iteration and output.

In the decryption process, the input data is IV (initialization vector, which we put into the buffer for hashing) and ciphertext. The output is plaintext. So the implementation would look like:

```c++
std::vector<uint8_t> mdBuffer;

// Put the IV (secret + auth) into the hash buffer
// ...

for (size_t i = 0; i < ciphertext.length() / 16; ++i)
{
    std::array<uint8_t, 16> md;
    // Make hash
    MD5(mdBuffer.data(), mdBuffer.size(), md.data());
    // XOR ciphertext with hash
    for (size_t j = 0; j < 16; ++j)
        plaintext[i * 16 + j] = ciphertext[i * 16 + j] ^ md[j];
    // Put the ciphertext block (secret + block) into the hash buffer
    // ...
}
```

In the encryption process, which you have below (`toVector`), the input data is IV and plaintext. The output is ciphertext. The cycle would look like this:
```c++
std::vector<uint8_t> mdBuffer;

// Put the IV (secret + auth) into the hash buffer
// ...

for (size_t i = 0; i < plaintext.length() / 16; ++i)
{
    std::array<uint8_t, 16> md;
    // Make hash
    MD5(mdBuffer.data(), mdBuffer.size(), md.data());
    // XOR ciphertext with hash
    for (size_t j = 0; j < 16; ++j)
        ciphertext[i * 16 + j] = plaintext[i * 16 + j] ^ md[j];
    // Put the ciphertext block (secret + block) into the hash buffer
    // ...
}
```